flake support

dguibert opened the pull request:
This introduces flake support to nixpack:
It exposes `lib.packs` and the spack packages behind pkgs attribute (bootstrapped with the default gcc compiler from nixpkgs).

````sh
nix build .#pkgs.mpileaks --impure --option sandbox false -L
````

As SPACK need to be patched, it is imported from flatironinstitute fork.
As nixpkgs need also to be patched [1], it is imported from my fork.

[1] https://github.com/NixOS/nixpkgs/pull/144747

dguibert committed:
nixpkgs: extract target as x86_64 -> x86-64

dguibert committed:
packs: allow to pass nixpkgs/spack directly instead of nixpkgsSrc/spackSrc

This will allow to pass these arguments from the flake inputs.

dguibert committed:
basic flake.nix support with lib.packs

impure and sandbox options should be passed as:

$ nix build .#pkgs.zlib -L --impure --option sandbox false

Qubasa commented:
@dguibert  Fails to build with `nix build .#pkgs.mpileaks --impure --option sandbox false -L ` 

```
$ nix-info
system: "x86_64-linux", multi-user?: yes, version: nix-env (Nix) 2.4, channels(mjnam): "", channels(harshanavkis): "home-manager-21.05", channels(root): "nix-ld, nixos-21.05.4448.4a8751d694b", nixpkgs: /nix/store/8hd2049cr4dz9myjcjjh6cjjapz1p107-source
``` 

```
Compiling Perl properties
Creating Perl synonyms
Writing tables
Making pod file
Making test script
Updating 'mktables.lst'
Traceback (most recent call last):
  File "/nix/store/vrr9dqpjss3mj1s8v8dkcpzv4111365j-builder.py", line 72, in <module>
    spack.installer.build_process(pkg, opts)
  File "/nix/store/k9lidw4pckdqls2yky5yrhnrah7klymq-source/lib/spack/spack/installer.py", line 1952, in build_process
    return installer.run()
  File "/nix/store/k9lidw4pckdqls2yky5yrhnrah7klymq-source/lib/spack/spack/installer.py", line 1809, in run
    self._real_install()
  File "/nix/store/k9lidw4pckdqls2yky5yrhnrah7klymq-source/lib/spack/spack/installer.py", line 1916, in _real_install
    phase(pkg.spec, pkg.prefix)
  File "/nix/store/vrr9dqpjss3mj1s8v8dkcpzv4111365j-builder.py", line 63, in wrapPhase
    return f(*args)
  File "/nix/store/k9lidw4pckdqls2yky5yrhnrah7klymq-source/lib/spack/spack/package.py", line 135, in phase_wrapper
    phase(spec, prefix)
  File "/nix/store/k9lidw4pckdqls2yky5yrhnrah7klymq-source/var/spack/repos/builtin/packages/perl/package.py", line 236, in build
    
  File "/nix/store/k9lidw4pckdqls2yky5yrhnrah7klymq-source/lib/spack/spack/build_environment.py", line 147, in __call__
    return super(MakeExecutable, self).__call__(*args, **kwargs)
  File "/nix/store/k9lidw4pckdqls2yky5yrhnrah7klymq-source/lib/spack/spack/util/executable.py", line 231, in __call__
    raise self.error
spack.util.executable.ProcessError: Command exited with status 2:
    'make' '-j16'
```

dguibert commented:
> @dguibert Fails to build with `nix build .#pkgs.mpileaks --impure --option sandbox false -L `

See commented Line 26 and Line 27 of flake.nix, PATH should contains some build tools as SPACK package.py are not as pure as NIX derivation.
Check the log of the failed derivation (nix log /nix/store/$derivation.drv)

As Dylan explained, the PATH variable can be populated with all the native tools of your OS, on NixOS I set
````nix
        PATH=/*"/run/current-system/sw/bin:"+*/
            inputs.nixpkgs.lib.concatStringsSep ":"
            (builtins.map (x: "${x}/bin")
            [
              pkgs.bash
              pkgs.coreutils
              pkgs.gnumake
              pkgs.gnutar
              pkgs.gzip
              pkgs.bzip2
              pkgs.xz
              pkgs.gawk
              pkgs.gnused
              pkgs.gnugrep
              pkgs.glib
              pkgs.binutils.bintools # glib: locale
              pkgs.patch
              pkgs.texinfo
            ]);
````

But perl does not build, I think the Perl package from SPACK hits this: https://github.com/Perl/perl5/pull/18752

Qubasa commented:
Ahh okay I see! I would suggest having a bit of documentation in the readme :-)

dguibert commented:
Hi @Luis-Hebendanz, I've just updated this PR
I can confirm that `nix build .#pkgs.mpileaks --impure --option sandbox false -L` works.

dguibert committed:
flake.lock: Add

Flake lock file changes:

• Added input 'nixpkgs':
    'github:NixOS/nixpkgs/dd956a5acab3b6ae370facb3066cf400d575ff44' (2022-04-29)
• Added input 'spack':
    'github:spack/spack/8a6b73bb2cb285904398224e743e405cea87dcce' (2022-04-29)

dguibert committed:
workaround nixpkgs 144747 (updated)

dguibert committed:
nixpkgs: libffi check fails with sssd
